### PR TITLE
feat: rewrite e2e test(test-e2e-route-with-method) using ginkgo

### PR DESCRIPTION
**Pull request by bisakhmondal**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**Related issues**
#1500

### CODE REVIEW

**Review by nic-chen — APPROVED** [review]
LGTM

File: api/test/e2enew/route/route_with_methods_test.go [hunk 1]
```diff
@@ -71,30 +75,27 @@ func TestRoute_with_methods(t *testing.T) {
...
 			ExpectStatus: http.StatusOK,
 			ExpectBody:   "hello world",
-			Sleep:        sleepTime,
-		},
-		{
-			Desc:         "delete route",
-			Object:       ManagerApiExpect(t),
+			Sleep:        base.SleepTime,
+		}),
+		table.Entry("delete route", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_with_methods_test.go:90]: It seems the update of these routes is missing, we should add it.

File: api/test/e2enew/route/route_with_methods_test.go [hunk 2]
```diff
@@ -110,73 +111,66 @@ func TestRoute_with_methods(t *testing.T) {
...
+			Headers:      map[string]string{"Authorization": base.GetToken()},
 			ExpectStatus: http.StatusOK,
 			ExpectBody:   "hello world",
-			Sleep:        sleepTime,
-		},
-		{
-			Desc:         "delete route",
-			Object:       ManagerApiExpect(t),
+			Sleep:        base.SleepTime,
+		}),
```
Comment (tokers) [inline api/test/e2enew/route/route_with_methods_test.go:164]: Also, we may update the `methods` list and verify the change.

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_methods_test.go:164]: looking into it.

**Comment by bisakhmondal** [comment]
Hello guys,

Can't we edit the URI of a route after it has been created? We can, right? 

It seems when we are editing/adding the existing/new URI nginx putting a 404 error, not the apisix.

it is how apisix put error for forbidden methods or non-existing URI
(before updating the URI)

image

But here (after editing)

image

Am I missing something here?
Thanks

**Comment by Jaycean** [comment]
> Hello guys,
> 
> Can't we edit the URI of a route after it has been created? We can, right?
> 
> It seems when we are editing/adding the existing/new URI nginx putting a 404 error, not the apisix.
> 
> it is how apisix put error for forbidden methods or non-existing URI
> (before updating the URI)
> 
> image
> 
> But here (after editing)
> 
> image
> 
> Am I missing something here?
> Thanks

I have just tested in my local area, and there is no 404 error, so I think you can create a new issue and write out the created data and process. Let's discuss this problem

**Comment by bisakhmondal** [comment]
Okay, @Jaycean, I'm opening an issue. 
See the ci is also failing with a 404 for that particular test.
image

**Review by tokers — APPROVED** [review]

**Comment by tokers** [comment]
@bisakhmondal Please fix the go lint errors :).

**Comment by bisakhmondal** [comment]
hi @tokers it's due to exceeding timeout, lol. Going to make a blank commit to retrigger the CIs

**Comment by tokers** [comment]
> hi @tokers it's due to exceeding timeout, lol. Going to make a blank commit to retrigger the CIs

I see, I'm triggering the CI actions.

**Comment by gxthrj** [comment]
E2E test is failed.

**Comment by Jaycean** [comment]
> Okay, @Jaycean, I'm opening an issue.
> See the ci is also failing with a 404 for that particular test.
> image

I see this problem. Let me find out the reason.

File: api/test/e2enew/route/route_with_methods_test.go [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["TEST"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:44]: nodes host please use the base.UpstreamIp

File: api/test/e2enew/route/route_with_methods_test.go [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["GET"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:72]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hellohello",
+					 "methods": ["GET"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:101]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["GET", "POST", "PUT", "DELETE", "PATCH"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:145]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["GET"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:213]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["GET", "post"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:265]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["GET"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:293]: ditto

File: api/test/e2enew/route/route_with_methods_test.go [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					 "name": "route1",
+					 "uri": "/hello",
+					 "methods": ["POST"],
+					 "upstream": {
+						 "type": "roundrobin",
+						 "nodes": [{
+							 "host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_methods_test.go:331]: ditto

File: api/test/e2enew/route/route_with_methods_test.go (hunk 3 above)
Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_methods_test.go]: Nice catch :+1:. Update pushed.

**Comment by Jaycean** [comment]
> Okay, @Jaycean, I'm opening an issue.
> See the ci is also failing with a 404 for that particular test.
> image

@bisakhmondal 
This error is that the URI created by APISIX is forwarded to the upstream, and the corresponding interface is also required. I checked the conf in the upstream image and found that the supporting interface is `hello` `hello1` `hello_` You can change update URI to `hello_`

File path：In the docker container:  johz/upstream
/usr/local/openresty/nginx/conf/lua/server.lua
```
function _M.hello()
    ngx.say("hello world")
end

function _M.hello1()
    ngx.say("hello1 world")
end

function _M.hello_()
    ngx.say("hello world")
end
```

**Comment by bisakhmondal** [comment]
Wonderful @Jaycean❤️. Thank you for the elaborated information.
So the whole time the 404 was thrown by upstream itself😶.

**Comment by Jaycean** [comment]
> Wonderful @Jaycean. Thank you for the elaborated information.
> So the whole time the 404 was thrown by upstream itself.

Yes, after all, this upstream is just for testing

**Comment by bisakhmondal** [comment]
> Yes, after all, this upstream is just for testing

No issues. Totally agree with you :)

**Review by Jaycean — APPROVED** [review]

**Review by juzhiyuan — APPROVED** [review]

**Pull request merged**